graphbolt/src/cuda/sampling_utils.cu

```diff
@@ -0,0 +1,132 @@
+}
+
+template <typename indptr_t, typename etype_t>
+struct EdgeTypeSearch {
+  const indptr_t* sub_indptr;
+  const indptr_t* sliced_indptr;
+  const etype_t* etypes;
+  int64_t num_fanouts;
+  int64_t num_rows;
+  indptr_t* new_sub_indptr;
+  indptr_t* new_sliced_indptr;
+  __host__ __device__ auto operator()(int64_t i) {
+    const auto homo_i = i / num_fanouts;
+    const auto indptr_i = sub_indptr[homo_i];
+    const auto degree = sub_indptr[homo_i + 1] - indptr_i;
+    const etype_t etype = i % num_fanouts;
+    auto offset = cuda::LowerBound(etypes + indptr_i, degree, etype);
+    new_sub_indptr[i] = indptr_i + offset;
+    new_sliced_indptr[i] = sliced_indptr[homo_i] + offset;
+    if (i == num_rows - 1) new_sub_indptr[num_rows] = indptr_i + degree;
+  }
+};
+
+std::tuple<torch::Tensor, torch::Tensor, torch::Tensor> SliceCSCIndptrHetero(
+    torch::Tensor sub_indptr, torch::Tensor etypes, torch::Tensor sliced_indptr,
+    int64_t num_fanouts) {
+  auto num_rows = (sub_indptr.size(0) - 1) * num_fanouts;
+  auto new_sub_indptr = torch::empty(num_rows + 1, sub_indptr.options());
+  auto new_indegree = torch::empty(num_rows + 2, sub_indptr.options());
+  auto new_sliced_indptr = torch::empty(num_rows, sliced_indptr.options());
+  auto allocator = cuda::GetAllocator();
+  auto stream = cuda::GetCurrentStream();
+  const auto exec_policy = thrust::cuda::par_nosync(allocator).on(stream);
+  thrust::counting_iterator<int64_t> iota(0);
+  AT_DISPATCH_INTEGRAL_TYPES(
+      sub_indptr.scalar_type(), "SliceCSCIndptrHeteroIndptr", ([&] {
+        using indptr_t = scalar_t;
+        AT_DISPATCH_INTEGRAL_TYPES(
+            etypes.scalar_type(), "SliceCSCIndptrHeteroTypePerEdge", ([&] {
+              using etype_t = scalar_t;
+              thrust::for_each(
+                  exec_policy, iota, iota + num_rows,
+                  EdgeTypeSearch<indptr_t, etype_t>{
+                      sub_indptr.data_ptr<indptr_t>(),
+                      sliced_indptr.data_ptr<indptr_t>(),
+                      etypes.data_ptr<etype_t>(), num_fanouts, num_rows,
+                      new_sub_indptr.data_ptr<indptr_t>(),
+                      new_sliced_indptr.data_ptr<indptr_t>()});
+            }));
+        size_t tmp_storage_size = 0;
+        cub::DeviceAdjacentDifference::SubtractLeftCopy(
+            nullptr, tmp_storage_size, new_sub_indptr.data_ptr<indptr_t>(),
+            new_indegree.data_ptr<indptr_t>(), num_rows + 1, cub::Difference{},
+            stream);
+        auto tmp_storage = allocator.AllocateStorage<char>(tmp_storage_size);
+        cub::DeviceAdjacentDifference::SubtractLeftCopy(
+            tmp_storage.get(), tmp_storage_size,
+            new_sub_indptr.data_ptr<indptr_t>(),
+            new_indegree.data_ptr<indptr_t>(), num_rows + 1, cub::Difference{},
+            stream);
+      }));
+  // Discard the first element of the SubtractLeftCopy result.
+  new_indegree = new_indegree.slice(0, 1, num_rows + 2);
```

Comment:
what's the magic number `2` for?

Reply:
equivalent to `[1:]` in python?

Reply:
Yes, I make sure to have the indegree tensor of size `num_rows + 1` so that doing an `ExclusiveCumSum` on it gives us the compacted `sub_indptr`.

Reply:
The first number is garbage due to AdjacentDifference not having a pair of values to subtract for the first value.

Reply:
does `new_indegree.slice(0, 1)` work?

Reply:
I will add comments about it.

Reply:
https://pytorch.org/cppdocs/api/function_namespaceat_1ae6bdafd2514da71bc5a656a649bcb575.html

Reply:
The arguments are dimension, start, and end. So that is the best I can do, unfortunately.

Reply:
Sorry, you are right. `.slice(0, 1)` should work.

Reply:
Yeah, it works. Fixed it. Is there anything else you want me to add here?